### DISCUSSION TITLE
Non-reconciled cells value handling

### DISCUSSION THREAD

**Discussion by malajvan** [DDMAL/linkedmusic-datalake#14] [General]
When handling cells without an entry on Wikidata, we have 2 choices of storing them before we decide to add them on Wikidata ourselves:
1. link them to another source (done in openrefine)
    This could be the databases' definition/where they store them, a dictionary, other trusted, long-lasting sources. The jsonld's key's value itself will stay as @id (as they already are since we're in the context of a reconciled column)

2. keep them as string literals
    We will keep the cell as-is, string values. To conform with jsonld format, we should change the type of the jsonld's key's value to @value instead of @id. You can do this by adding extra context in each JSON document to override the @context if needed.  

You can see the examples with cantusdb's genres column (P136). Most of the values are reconciled with an existing Q on Wikidata, but a handful are not.

1. link them to another source: [link_compact.jsonld](https://github.com/malajvan/linkedmusic-datalake/blob/main/cantusdb/jsonld/link_compact.jsonld)
- the first document's P136 is a wikidata id, while the second is a link to specific genre in the cantusdb website
```
[
    {
        "@context": [
            "https://raw.githubusercontent.com/malajvan/linkedmusic-datalake/main/cantusdb/jsonld/context.jsonld"
        ],
        "@id": "chant:643187",
        "@type": "wd:Q23072435",
        "P136": "wd:Q224377",
......
    },
    {
        "@context": [
            "https://raw.githubusercontent.com/malajvan/linkedmusic-datalake/main/cantusdb/jsonld/context.jsonld"
        ],
        "@id": "chant:638102",
        "@type": "wd:Q23072435",
        "P136": "https://cantusdatabase.org/genre/134",
.....
    }
```

- since the context didn't change, the expanded version of these 2 will look similar. (shortened)
```
[
  { ......
    "http://www.wikidata.org/prop/direct/P136": [
      {
        "@id": "http://www.wikidata.org/entity/Q224377"
      }
    ],
   ........

    "http://www.wikidata.org/prop/direct/P136": [
      {
        "@id": "https://cantusdatabase.org/genre/134"
      }
    ]
  }
]
```

2. keep them as string literals ([val_compact.json](https://github.com/malajvan/linkedmusic-datalake/blob/main/cantusdb/jsonld/val_compact.jsonld)

- same looking at P136, we have, for the same chants:
```
    {
        "@context": [
            "https://raw.githubusercontent.com/malajvan/linkedmusic-datalake/main/cantusdb/jsonld/context.jsonld"
        ],
        "@id": "chant:643187",
        "@type": "wd:Q23072435",
        "P136": "wd:Q224377",
....
    },
    {
        "@context": [
            "https://raw.githubusercontent.com/malajvan/linkedmusic-datalake/main/cantusdb/jsonld/context.jsonld",
            {
                "P136": "wdt:P136"
            }
        ],
        "@id": "chant:638102",
        "@type": "wd:Q23072435",
        "P136": "Dramatic element (used for items of liturgical drama that are not otherwise rubricked)",
....
    }
```

notice that on the @context of the second chant, we added "P136": "wdt:P136", this overrides the P136 definition in the context link above. This will expand to this:

```
[
  { ......
    "http://www.wikidata.org/prop/direct/P136": [
      {
        "@id": "http://www.wikidata.org/entity/Q224377"
      }
    ],
   ........

    "http://www.wikidata.org/prop/direct/P136": [
      {
       "@value": "Dramatic element (used for items of liturgical drama that are not otherwise rubricked)"
      }
    ]
  }
]
```

I'm unsure which one would be better for keeping the integrity of the data as well as facilitating searching and indexing.

**Comment by dchiller** [comment]
> notice that on the @context of the second chant, we added "P136": "wdt:P136", this overrides the P136 definition in the context link above. 

To confirm my understanding of option 2, with this override we are essentially removing `"@type" = "@id"` line from the context:
```
        "P136": {
            "@id": "wdt:P136",
            "@type": "@id"
        }
```

so that then when this is expanded we know that the "value" of the "P136" field is not an IRI/"@id" but is just has a string literal "@value".

Is that right?

**Reply by malajvan** [reply]
That’s correct

**Comment by dchiller** [comment]
> I'm unsure which one would be better for keeping the integrity of the data as well as facilitating searching and indexing.

Ok one thought I'm having for the searching is let's say I want to find all of the Introits and all of the Dramatic Elements. For something that we **can** reconcile to a wikidata item (like Introits) then when someone wants to search for that thing we can go find the Wikidata IRI and search our indexed data for that IRI. But for something we **can't** reconcile to wikidata (like a dramatic element), if we've indexed a non-wikidata IRI (eg. option 1) then we have to figure out what the IRI is, what database it's from, etc, which might be hard if we start doing this with a bunch of different databases. Whereas if we've indexed just the value ("Dramatic element....") then we can search our indexed data for that string directly.  

Not sure that makes any sense?

**Reply by malajvan** [reply]
Yes that makes sense, and I agree with this approach too, if the @value and @id difference doesn’t affect the search.
